soc/arm/silabs_sim3/sim3u/soc.c

```diff
@@ -0,0 +1,128 @@
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_0, PORTBANK_0_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_0, PORTBANK_0_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_0, PORTBANK_0_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_0, PORTBANK_0_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_1, PORTBANK_1_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_1, PORTBANK_1_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_1, PORTBANK_1_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_1, PORTBANK_1_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_1, PORTBANK_1_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_2, PORTBANK_2_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_2, PORTBANK_2_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_2, PORTBANK_2_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_2, PORTBANK_2_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_2, PORTBANK_2_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_3, PORTBANK_3_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_3, PORTBANK_3_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_3, PORTBANK_3_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_3, PORTBANK_3_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_3, PORTBANK_3_PINS_ANALOG);
+
+	/* Configure skips */
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_0, PORTBANK_0_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_1, PORTBANK_1_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_2, PORTBANK_2_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_3, PORTBANK_3_SKIPEN_VALUE);
+
+	/* Configure crossbars */
+	SI32_PBCFG_A_enable_xbar0l_peripherals(SI32_PBCFG_0, CROSSBAR_0_CONFIG & 0xFFFFFFFF);
+	SI32_PBCFG_A_enable_xbar0h_peripherals(SI32_PBCFG_0,
+					       (CROSSBAR_0_CONFIG >> 32) & 0xFFFFFFFF);
+	SI32_PBCFG_A_enable_xbar1_peripherals(SI32_PBCFG_0, CROSSBAR_1_CONFIG & 0xFFFFFFFF);
+
+	/* Enable crossbars */
+	SI32_PBCFG_A_enable_crossbar_0(SI32_PBCFG_0);
+	SI32_PBCFG_A_enable_crossbar_1(SI32_PBCFG_0);
+}
+
+static void vddlow_irq_handler(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* nothing to do here, we just don't want any spurious interrupts */
+}
+
+static void vmon_init(void)
+{
+	/* VMON must be enabled for flash write/erase support */
+
+	NVIC_ClearPendingIRQ(VDDLOW_IRQn);
+
+	IRQ_CONNECT(VDDLOW_IRQn, 0, vddlow_irq_handler, NULL, 0);
+	irq_enable(VDDLOW_IRQn);
+
+	SI32_VMON_A_enable_vdd_supply_monitor(SI32_VMON_0);
+	SI32_VMON_A_enable_vdd_low_interrupt(SI32_VMON_0);
+	SI32_VMON_A_select_vdd_standard_threshold(SI32_VMON_0);
+}
+
+__no_optimization static void busy_delay(uint32_t cycles)
+{
+	while (cycles) {
+		cycles--;
+	}
+}
+
+static int silabs_sim3u_init(void)
+{
+	uint32_t key;
+
+	key = irq_lock();
+
+	/* The watchdog may be enabled already so we have to disable it */
```

Comment:
Enabled by whom - is there some kind of bootrom on this SoC that enables some of the peripherals? What's the default reset state for watchdog on SIM3U?

Reply:
Unfortunately I don't know that. It's just something that I have observed to be the case on certain devices.